added gpu test script and associated aws infrastructure code

RobJY opened the pull request:
Just submitting this PR to start the conversation about adding gpu testing infrastructure to the repo.

I'm submitting a bash script that tests for some basic gpu functionality and have included code I use to spin up an AWS gpu enabled instance for testing.
  
PR contains the following files:
tests/gpu/test-gpu.sh - bash script that checks for gpu functionality like: valid gpu driver, gpu toolkit, CUBLAS and tensorflow.
tests/gpu/misc - contains code I use for spinning up a rocker container on a gpu enabled AWS instance:
tests/gpu/misc/deploy-aws-rocker-ml-gpu.yml - Ansible playbook that installs CUDA driver on instance and builds Rocker container with Rocker image, CUDA samples & tensorflow installed.
tests/gpu/misc/Dockerfile - Dockerfile for build
tests/gpu/misc/docker-compose.yml - docker-compose file for building and running container with gpu support
tests/gpu/misc/hosts - hosts file for Ansible playbook

There are lots more applications in the CUDA samples that could be included in the test script, so please let me know if any seem helpful. 

I was planning to build this out to test for more drivers and toolkit versions and eventually more cards if that sounds good.  Please let me know if you have ideas about how I should proceed.

Please let me know any thoughts or suggestions.  Thanks!

Cheers,
Rob

cboettig commented:
Thanks for starting the conversation on this.  

Just a heads up that it looks like you may be building on some outdated rocker images though.  Current GPU images are `rocker/ml` and `rocker/ml-verse`, both available with `cuda10.1` and `cuda11.1` tags , along with any R 4.x version tag.  See README over in:
https://github.com/rocker-org/ml, though that's just there for historical / docker hub reasons, as all sources are now here in versioned2.

You'll find the scripts currently used to do the CUDA installs; though cuda10.1 stack takes the strategy of adding cuda10.1 onto rocker based images, while the 11.1 stack uses NVIDIA's ubuntu:focal based `cuda11.1` as the base.  (NVIDIA doesn't have cuda10.1-base images available on ubuntu-focal, only on older LTSs).  Feedback on the existing stack would be more than welcome!  


I'll try and take a closer look at what you have here as well, but wanted to get us on the same page; looks like this may be building on `rocker/ml-gpu` which is stuck on cuda9 and not maintained.

RobJY commented:
Thanks for the heads up about me using the wrong rocker images!  I think I've got it right now.  I'm building separate 10 and 11 containers with the cuda samples and testing code built in.  I'm currently running the tests from the host with the commands:
`sudo docker exec container-id /test-gpu.sh 10`
or
`sudo docker exec container-id /test-gpu.sh 11`
depending on the cuda version in the rocker container.

RobJY commented:
Hmm, trying to do deeper testing of tensorflow and I'm getting errors from each container.  Looks like some missing libraries.  Working on a fix.

cboettig commented:
@RobJY thanks for the update and keep me posted.  Tensorflow is particularly picky about cuda lib versions so definitely could be an issue.  Can you report the version of tensorflow you're using and the version of the cuda libs it wants?


I'm mostly using these images with pytorch workflows so the Tensorflow side is less tested.  We have some legacy stuff that needs Tensorflow 1.x, which is a huge pain since it's not available for python 3.8 and we have to install an older python version to get it working, we've been using pipenv+pyenv to handle python version switching.  See https://github.com/rocker-org/ml#python-versions-and-virtualenvs.  

For Tensorflow  1.x, I recall it requests the 10.0 versions of cuda libs, not the 10.1 versions we ship with.  I've found it works to monkey-patch these in like so: https://github.com/rocker-org/rocker-versioned2/blob/master/scripts/install_tf1_cuda_10_0.sh.  If you have a different version of tensorflow you probably need a different set of these. (Fortunately I think the apt repos we already added from NVIDIA should have all versions available).  We should probably figure out if there's tensorflow binaries avialable that link against cuda 10.1 and cuda 11.1, and pre-install that version of tensorflow for the user.  


Anyway, thanks again for the test notes and keep it coming.

RobJY commented:
Thanks for the cuda tensorflow info!  This looks really helpful.  The 10.1 version seems more fussy than the 11.1, so I may start with 11.1.  My original test was just a check of tensorflow package installation and not any actual gpu computation.  It's when I started to add gpu tests that I got the library errors.

Noam had said that I should test a range of tensorflow versions because it was so picky.  My plan is to start by getting a simple gpu tensorflow test to work for each version and then build in version testing.  I was thinking I'd start with a low version of tensorflow and then upgrade it through versions while running tests and then the script can report back which ones had problems, if any.

Please let me know if you have any other suggestions or thoughts.

RobJY commented:
In those last commits I think I've got this working for cuda 11.  Currently testing tensorflow versions 1.15, 2.2 and latest (~2.6).  I'll look up the specific versions of 2 you mention above and add them then start on cuda 10.

RobJY commented:
Noam has pointed out two important issues:

- I need to be using the same tensorflow install mechanisms in my testing code that Rocker uses.  Noam pointed me to lots of code so I'll take a look and put together a plan.
- I need to add tests at the R level.  I currently test at hardware, OS and Python, but not R, so I'll add that to the cuda11 version before proceeding to cuda10.

RobJY commented:
I'm having trouble running rocker-versioned2/tests/ml/nvblas.R in my cuda.11 container. When I run:
`NVBLAS_CONFIG_FILE=/etc/nvblas.conf LD_PRELOAD=/usr/local/cuda-11.1/lib64/libnvblas.so.11 R`
I get the error:
```Error in get_result(output = out, options) : 
  callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Type .Last.error.trace to see where the error occurred```
Here's the trace:
```
 Stack trace:

 1. callr::r(function() { ...
 2. callr:::get_result(output = out, options)
 3. throw(new_callr_error(output, killmsg))

 x callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed 
```
I tracked this to /var/log/nvblas.log:
```
[NVBLAS] Cannot open CPU Library '/usr/lib/x86_64-linux-gnu/openblas/libblas.so.3'
[NVBLAS] Using devices :0 
[NVBLAS] Config parsed
[NVBLAS] CPU Blas library need to be provided
```
and added NVBLAS_CPU_BLAS_LIB env var to R startup:
`NVBLAS_CONFIG_FILE=/etc/nvblas.conf LD_PRELOAD=/usr/local/cuda-11.1/lib64/libnvblas.so.11 NVBLAS_CPU_BLAS_LIB=/usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 R`
but that didn't change anything.  I was able to hack a fix by making a symlink from /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3 to /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3, so it doesn't seem to be taking that environment variable value.  Any idea why that would be?

cboettig commented:
@RobJY The NVBlas configuration is done here: https://github.com/rocker-org/rocker-versioned2/blob/master/scripts/config_R_cuda.sh

Looks like it has the wrong paths, at least for CUDA 11.  Note that these need to also go in `/etc/nvblas.conf`, see

https://github.com/rocker-org/rocker-versioned2/blob/b8c5065da66d57cf549c05c4b2836b4af466d7f1/scripts/config_R_cuda.sh#L28-L32

As you see, the symlink trick worked probably because that way it matched what we already had put in `/etc/nvblas.conf`.  The install already creates a bunch of symlinks I think(? haven't looked at this in years so I could be way out of date) which ought to make it easier to link against without hardcoding precise versions into `/etc/nvblas.conf`

Thanks for testing this.  nvblas is one of those things that sounds like a cool idea but is really quiet a nuisance too. Kind of severe that it decides to crash instead of just falling back on some more standard blas when it can't find the library...

RobJY commented:
I've got a version of the testing script and Docker file working with four versions of tensorflow (1.15.5, 2.2, 2.5, 2.6), but to get version 1 of tensorflow installed I need to use:
```
RUN pip install nvidia-pyindex
RUN pip install nvidia-tensorflow[horovod]
```
If I don't I get the error:
```
ERROR: Could not find a version that satisfies the requirement tensorflow-gpu==1.14.0 (from versions: 2.2.0rc1, 2.2.0rc2, 2.2.0rc3, 2.2.0rc4, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.3.0rc0, 2.3.0rc1, 2.3.0rc2, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.4, 2.4.0rc0, 2.4.0rc1, 2.4.0rc2, 2.4.0rc3, 2.4.0rc4, 2.4.0, 2.4.1, 2.4.2, 2.4.3, 2.5.0rc0, 2.5.0rc1, 2.5.0rc2, 2.5.0rc3, 2.5.0, 2.5.1, 2.6.0rc0, 2.6.0rc1, 2.6.0rc2, 2.6.0)
ERROR: No matching distribution found for tensorflow-gpu==1.14.0
WARNING: You are using pip version 21.1.2; however, version 21.2.4 is available.
You should consider upgrading via the '/opt/venv/reticulate/bin/python -m pip install --upgrade pip' command.
Error: Error installing package(s): 'tensorflow-gpu==1.14.0', 'tensorflow-probability==0.7.0'
Execution halted
Failed to install tensorflow version 1.14.0
```

when I install using:
```
    install.packages("tensorflow")
    tensorflow::install_tensorflow(version="${ver}-gpu", extra_packages="tensorflow-probability==0.7.0")
```
as shown in [this page](https://hub.docker.com/r/rocker/cuda)

This is just for cuda version 11.  I haven't tried version 10 yet.

Should I be installing tensorflow with a different command?

RobJY commented:
Same for cuda10 version.
I had to downgrade gcc to get the cuda samples to compile, but no difference with regards to tensorflow.

RobJY commented:
Current plan:

1. Ansible playbook that will install different versions of nvidia driver and then run bash test script.  (started before I went on vacation)
2. clean up bash test script output.  I think it would be best to dump verbose output to a log file and just report a brief summary to standard out.
3. look at existing config files with eye to integrating test code.  I think the bash script or parts of it could be integrated, but the driver tests will require removal of existing driver and reboot, so it will probably need to be a stand alone test.

Please let me know if you'd like any adjustment to this plan or have any other ideas.

cboettig commented:
@RobJY nice! that sounds like a good plan to me.

RobJY commented:
the last commits clean up the test-gpu.sh script output and send verbose messages to a log file.  I also added test-gpu-outside.sh that can be run from outside the containers and dumps an output log file to /gpu-test.log on the host.  This script is run by the Ansible playbook test-drivers.yml.  It runs test-gpu.sh with the 460 and 470 driver.  I tried it on older drivers, but it always defaulted to the 460.  I guess the 460 is backward compatible with the older ones at least for Ubuntu 20.04.
Warning, the test-drivers.yml playbook is dangerous in that it uninstalls all packages on the host with the 'nvidia-driver' prefix.  

Please let me know if you'd like any changes.

I'm looking at possible changes to config_R_cuda.sh now.

RobJY commented:
The last commit adds proposed tests to scripts/config_R_cuda.sh.  I just added some of the simplest tests to check driver, toolbox and tensorflow/nvblas installation.  Errors are echo'ed and sent to nvblas.log file.  Please let me know if this is acceptable or if you'd like changes.

I also discovered a couple issues while testing.  The first was that I was clobbering the test-gpu.sh log file, so I fixed that.

Another issue that I don't fully understand yet is that I changed the path in config_R_cuda.sh for NVBLAS_CPU_BLAS_LIB to /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 as we discussed earlier.  I then removed the /etc/nvblas.conf file copy in my test Dockerfiles since it accomplishes the same thing.  After I made this change the cuda version 10 container tests ran fine, but the cuda version 11 container tests failed with the same message mentioned above.  Is config_R_cuda.sh not being run for rocker/ml:4.0.5-cuda11.1?  I'll try to look into this more.

Current status is that I'm still copying over the updated nvblas.conf file for the cuda 11 version Dockerfile, so it will work.

cboettig commented:
@RobJY -- in addition to installing tensorflow, we will want to make sure the tests run tensorflow's `is_gpu_available` test, https://www.tensorflow.org/api_docs/python/tf/test/is_gpu_available for each cuda arch on each version of tensorflow.  This will likely surface more issues with cudnn versions etc as TF is very particular about libraries.  

The above changes sound reasonable though I haven't had too much time to dig in. Thanks for catching the issue with the log file.  


Weird. `config_R_cuda.sh` should be being called on the 11.1 Dockerfile, it's in the stack definition here: https://github.com/rocker-org/rocker-versioned2/blob/master/stacks/4.1.1.json#L317, and you can see it in the corresponding parent Dockerfile for cuda11.1 here: https://github.com/rocker-org/rocker-versioned2/blob/master/dockerfiles/cuda11_4.1.1.Dockerfile#L28.  `config_R_cuda.sh` was never fully modified/tested to make sure it had the paths appropriate to the upstream NVIDIA cuda 11.1 Dockerfiles though, so maybe it needs patching for that?

eitsupi commented on tests/gpu/misc/Dockerfile.cuda10:
```diff
@@ -0,0 +1,23 @@
+RUN make
+
+COPY ./test-gpu.sh /test-gpu.sh
+RUN chmod +x /test-gpu.sh
```
This line may not be needed if you run `chmod +x test-gpu.sh` to make `test-gpu.sh` executable before committing.

RobJY commented:
Latest commit adds Tensorflow code from examples.R to test Tensorflow versions.  Log output is messy, but results look like:

- driver 460
   - toolbox 10
      - tf 1.15: works
      - tf 2.2: works
      - tf 2.5: fails
      - tf 2.6: fails
   - toolbox 11
      - tf 1.15: works
      - tf 2.2: fails
      - tf 2.5: works
      - tf 2.6: works
- driver 470
   - toolbox 10
      - tf 1.15: fails
      - tf 2.2: works
      - tf 2.5: fails
      - tf 2.6: fails
   - toolbox 11
      - tf 1.15: fails
      - tf 2.2: fails
      - tf 2.5: fails
      - tf 2.6: works
I'll try cleaning up the log messages as well as looking in to the causes of the failures, starting with the most recent driver and toolbox.

Thank you for the suggestion @eitsupi, I'll make that change.

RobJY commented:
As I dig into the errors from the above combinations I'm finding that my flag for success needs to be updated.  That means that more combinations work than I originally thought.  I'll keep updating the results and commit the updated script when I'm done.

RobJY commented:
Okay, I think the test script is now correct and the logs don't get clobbered when run from the Ansible playbook test-drivers.yml.

RobJY commented:
I think the last commit completes the stand-alone scripts for gpu testing.  Please let me know if you see any issues or would like me to add or change anything.  Maybe I should add a warning to the start of the test-drivers.yml playbook since it uninstalls nvidia-driver-* from the system.  What do you think?

I can make changes to the config and build functions as we discussed, but let's put those things under separate pull requests.  How does that sound?

RobJY commented:
Ah, looking over the ticket I'm reminded of the issue I had with the config file in the cuda 11 container, but not the cuda 10 one.  I'll look into that a bit more.

RobJY commented:
Yeah, I'm still getting the error when I remove my copy over of /etc/nvblas.conf.  I see where you say config_R_cuda.sh is being run, but I'm not sure it's actually getting run for the cuda 11 container.  I don't find /var/log/nvblas.log in my cuda 11 container which should have been created when config_R_cuda.sh was run.  I'll keep looking into this.

RobJY commented:
Sorry, please disregard the part above about not seeing /var/log/nvblas.log.  I'm finding that now in the cuda 11 container.  I'm not sure why I didn't see it previously.  Maybe I was looking on the host and not the container.

I'm still getting the error mentioned above for the cuda 11 container, so I'll look into fixing that.

RobJY commented:
I've started with a clean rebuild of both containers and I no longer see updates to /etc/nvblas.conf, which I think is correct because config_R_cuda.sh only gets applied as part of your build process.  I must have had an old updated image that was confusing me.

I believe you're correct that all that needs to be done is to patch the existing version of config_R_cuda.sh with the version I commited and the errors for nvblas will be fixed.  I'll continue to add the line in my testing Dockerfiles until that change has been made.

Now I'll see if using is_gpu_available gets the same results as my current tensorflow test code and replace it if so, since that would be quicker. Although it might be nice to have the longer test code in the stand-alone test script for benchmarking.  I could then just use is_gpu_available in the shorter tests that run as part of the config script.

RobJY commented:
I tried using `is_gpu_available()`, but it passes for all versions of tensorflow/toolkit/driver.  This makes sense because it's run from python, so it's not really testing how R interacts with the tensorflow/toolkit/driver versions.  The script I've been using is the tensorflow section from your examples.R file, so it is testing how R works with all the versions.  I think we should stick with this approach.

I think the code in this PR is good to merge, so would you review it when you have a chance and let me know if you'd like any changes or see any problems?

While we're closing this PR I can make a new one to investigate the errors found by the test script and try to correct them.  I can also open another PR to discuss adding tests to config_R_cuda.sh.  I had a version of this in an earlier commit on this PR, but removed it because I think we'll need more discussion about it.

Does this sound acceptable?

cboettig commented:
Thanks this sounds great!   Reviewing this one now, and :+1: to the two separate PR proposals you suggest above for addressing errors and for discussing additional tests for `config_R_cuda.sh`

cboettig reviewed (commented):
looks good, added a few comments here.

RobJY commented on scripts/config_R_cuda.sh:
```diff
@@ -26,7 +26,7 @@ chmod a+rw /var/log/nvblas.log
 ## Allow R to use CUDA for BLAS, with fallback on openblas
 echo 'NVBLAS_LOGFILE /var/log/nvblas.log
-NVBLAS_CPU_BLAS_LIB /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
+NVBLAS_CPU_BLAS_LIB /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
```
Sure, sounds like a great idea.  Something like the following on line 28?
## NOTE: NVBLAS_CPU_BLAS_LIB must be correct for UBUNTU_VERSION selected in scripts/install_R.sh#L25

RobJY replied:
I've added the comment above as suggested.  Please let me know if you'd like any changes.

RobJY commented on tests/gpu/misc/Dockerfile.cuda10:
```diff
@@ -0,0 +1,29 @@
+FROM rocker/ml:4.0.5-cuda10.1
```
Yeah, sounds good.  Let me try building with that change.

RobJY commented on tests/gpu/misc/examples_tf.R:
```diff
@@ -0,0 +1,36 @@
+## Tensorflow:
+install.packages('keras', repos='http://cran.us.r-project.org')
+library(keras)
+mnist <- dataset_mnist()
```
You're correct.  It attempts to fall back to CPU.  I catch GPU runs by looking in the output for it to define a GPU device like so:
```
    GPU_STR="device:GPU:0"
    if [[ "$TF_OUTPUT" == *"$GPU_STR"* ]]
```
Please let me know if there's a better way to test for GPU success.

RobJY replied:
I think this is good to go unless you'd like me to make a change.

RobJY commented on tests/gpu/misc/nvblas.conf:
```diff
@@ -0,0 +1,3 @@
+NVBLAS_LOGFILE /var/log/nvblas.log
```
I won't need it once the change to config_R_cuda.sh makes it into the build process, but will until then because I'm pulling the built containers.
I can remove it from the PR though to reduce clutter.

RobJY replied:
It's actually a little difficult to remove it before the change to config_R_cuda.sh has been deployed.  I would have to remove it from my Ansible and Docker code as well.  After this has been deployed I can easily remove all references to it.  Would that be okay?

RobJY commented:
The last commits fix the shellcheck errors and address the 4 conversations started above.  I'll respond to each of those conversations above with details.

cboettig commented:
:tada: 

Thanks @RobJY

RobJY commented:
I tested building the containers and running the gpu tests with nvblas.conf removed from tests/gpu/misc and the nvblas tests worked fine.  I just wanted to make sure that was good since we moved it to config_R_cuda.sh.  I'll make a little pull request to merge that clean up in before I get started on the other PRs we discussed.

cboettig commented:
:clap: :+1: Thanks @RobJY , keep me posted.
